polygon/bor/bordb/prune.go

```diff
@@ -17,27 +17,321 @@
 package bordb
 
 import (
+	"context"
 	"encoding/binary"
-	"errors"
+	"fmt"
 
 	"github.com/erigontech/erigon-lib/kv"
-	"github.com/erigontech/erigon/polygon/bor/snaptype"
+	"github.com/erigontech/erigon/polygon/bridge"
 	"github.com/erigontech/erigon/polygon/heimdall"
 )
 
+type HeimdallUnwindCfg struct {
+	KeepEvents                      bool
+	KeepEventNums                   bool
+	KeepEventProcessedBlocks        bool
+	KeepSpans                       bool
+	KeepSpanBlockProducerSelections bool
+	KeepCheckpoints                 bool
+	KeepMilestones                  bool
+	Astrid                          bool
+}
+
+func (cfg *HeimdallUnwindCfg) ApplyUserUnwindTypeOverrides(userUnwindTypeOverrides []string) {
+	if len(userUnwindTypeOverrides) > 0 {
+		return
+	}
+
+	// If a user has specified an unwind type override it means we need to unwind all the tables that fall
+	// inside that type but NOT unwind the tables for the types that have not been specified in the overrides.
+	// Our default config value unwinds everything.
+	// If we initialise that and keep track of all the "unseen" unwind type overrides then we can flip our config
+	// to not unwind the tables for the "unseen" types.
+	const events = "events"
+	const spans = "spans"
+	const checkpoints = "checkpoints"
+	const milestones = "milestones"
+	unwindTypes := map[string]struct{}{
+		events:      {},
+		spans:       {},
+		checkpoints: {},
+		milestones:  {},
+	}
+
+	for _, unwindType := range userUnwindTypeOverrides {
+		if _, exists := unwindTypes[unwindType]; !exists {
+			panic("unknown unwindType override " + unwindType)
+		}
+
+		delete(unwindTypes, unwindType)
+	}
+
+	// our config unwinds everything by default
+	defaultCfg := HeimdallUnwindCfg{}
+	defaultCfg.Astrid = cfg.Astrid
+	// flip the config for the unseen type overrides
+	for unwindType := range unwindTypes {
+		switch unwindType {
+		case events:
+			defaultCfg.KeepEvents = true
+			defaultCfg.KeepEventNums = true
+			defaultCfg.KeepEventProcessedBlocks = true
+		case spans:
+			defaultCfg.KeepSpans = true
+			defaultCfg.KeepSpanBlockProducerSelections = true
+		case checkpoints:
+			defaultCfg.KeepCheckpoints = true
+		case milestones:
+			defaultCfg.KeepMilestones = true
+		default:
+			panic(fmt.Sprintf("missing override logic for unwindType %s, please add it", unwindType))
+		}
+	}
+
+	*cfg = defaultCfg
+}
+
+func UnwindHeimdall(ctx context.Context, heimdallStore heimdall.Store, bridgeStore bridge.Store, tx kv.RwTx, unwindPoint uint64, unwindCfg HeimdallUnwindCfg) error {
+	if !unwindCfg.KeepEvents {
+		if err := UnwindEvents(tx, unwindPoint); err != nil {
+			return err
+		}
+	}
+
+	if !unwindCfg.KeepEventNums {
+		if err := UnwindEventNums(tx, unwindPoint); err != nil {
```

Comment:
on main branch we use `bridge.UnwindBlockNumToEventID` here (so can delete local UnwindEventNums func)
same with function in the next if we use `bridge.UnwindEventProcessedBlocks` here (so can delete local UnwindEventProcessedBlocks func)
probably left over from a merge

Reply:
I think that this:

```go
if !unwindCfg.KeepEvents {
	if err := UnwindEvents(tx, unwindPoint); err != nil {
		return err
	}
}
```

Can become bridge,UnwindEvents - however it seems to contain the KeepEventNums/KeepEventProcessedBlocks so I'm not sure how it relates to the code in the local function.